QuickEditor: Refreshing all info when no avatar selected and a new one is uploaded

hamorillo opened the pull request:
Closes #380 

### Description

When a new avatar is uploaded into a Gravatar account with no avatar selected, the backend automatically uses that image as the default avatar.

We need to reflect that in the QE UI. We can hardcode the changes in the code, modifying the uiState as required. However, if the backend behavior changes at some point, all previous versions of the QE will show the wrong status of the Gravatar account.

To mitigate that situation, we will refresh all data from the backend if an image is uploaded and no previous avatar is selected.

What do you think about this approach? Do you think it's worth it to refresh everything or assume the risk and modify the uiState locally, expecting the backend to do the same?

**Note:** I'll add some unit tests before merging if we decide to go this way.

### Testing Steps

1. Using an account with no avatar selected
2. Upload a new image using the QE
3. Verify you see, after reloading, the new image as the selected image.

hamorillo committed:
Refreshing all info when no avatar selected and a new one is uploaded

When a new avatar is uploaded into a Gravatar account with no avatar selected, the backend automatically uses that image as the default avatar.

We need to reflect that in the QE UI. We can hardcode the changes in the code, modifying the uiState as required. However, if the backend behavior changes at some point, all previous versions of the QE will show the wrong status of the Gravatar account.

To mitigate that situation, if an image is uploaded and no previous avatar is selected, we will refresh all data from the backend.

AdamGrzybkowski commented on gravatar-quickeditor/src/main/java/com/gravatar/quickeditor/ui/avatarpicker/AvatarPickerViewModel.kt:
```diff
@@ -146,18 +146,26 @@ internal class AvatarPickerViewModel(
-                        )
+                    val isAutoSelected = _uiState.value.emailAvatars?.selectedAvatarId == null
+                    if (isAutoSelected) {
+                        _uiState.update { AvatarPickerUiState(email, avatarPickerContentLayout) }
+                        refresh()
```
I think it's completely reasonable to refresh the content like you said, to future proof in case of backend changes, but I wonder if we really have to clean the whole state?

Could we do this refresh action as part of the uploading indicator? With the current code, we hide the whole list, the profile card is recomposed fully and that seems like a lot to select an avatar.

hamorillo replied:
> Could we do this refresh action as part of the uploading indicator? With the current code, we hide the whole list, the profile card is recomposed fully and that seems like a lot to select an avatar.

I don't think it's easy to keep the uploading indicator until the `fetchAvatars` finishes, at least reusing the current methods, but I'll try to improve it.

AdamGrzybkowski replied:
Correct, you have to use suspend functions that won't start a new coroutine. We do have one like that to [fetch avatars](https://github.com/Automattic/Gravatar-SDK-Android/blob/trunk/gravatar-quickeditor/src/main/java/com/gravatar/quickeditor/ui/avatarpicker/AvatarPickerViewModel.kt#L209) but not for the profile.

hamorillo replied:
Done!

hamorillo committed:
Keep loading the state while refreshing avatars after uploading a new one

Currently, in an account with no avatar selected, when a new avatar is uploaded, the backend auto-selects it. We should try to keep that as transparent as possible in the UI for the final user, so we want to keep the uploading state until the avatars are fetched again.

hamorillo commented on gravatar-quickeditor/src/main/java/com/gravatar/quickeditor/ui/avatarpicker/AvatarPickerViewModel.kt:
```diff
@@ -217,9 +230,9 @@ internal class AvatarPickerViewModel(
                         emailAvatars = emailAvatars,
                         scrollToIndex = if (scrollToSelected && emailAvatars.avatars.isNotEmpty()) {
-                            emailAvatars.avatars.indexOfFirst {
+                            emailAvatars.avatars.indexOfFirstOrNull {
```
With the `coerceAtLeast(0)`, when there wasn't an avatar selected, we still updated the state with 0, so when we uploaded a new image and set it to 0 again, the `scrollToIndex` there wasn't a change in the state, and nothing happens. I think this is more correct and solves the issue.

AdamGrzybkowski commented on gravatar-quickeditor/src/main/java/com/gravatar/quickeditor/ui/avatarpicker/AvatarPickerViewModel.kt:
```diff
@@ -146,18 +146,31 @@ internal class AvatarPickerViewModel(
-                            scrollToIndex = null,
-                        )
+                    val isAutoSelected = _uiState.value.emailAvatars?.selectedAvatarId == null
+                    if (isAutoSelected) {
+                        fetchAvatars(showLoading = false, scrollToSelected = true)
+                        _uiState.update { currentState ->
+                            currentState.copy(
+                                uploadingAvatar = null,
+                                avatarUpdates = currentState.avatarUpdates.inc(),
+                            )
+                        }
+                    } else {
```
Shouldn't we emit the `AvatarPickerAction.AvatarSelected` here?

hamorillo replied:
Good catch! Done!

hamorillo committed:
Send AvatarSelected action when the avatar is autoSelected by the backend

If the backend has autoselected the avatar, we need to send the AvatarSelected action as with any other avatar selection.

AdamGrzybkowski reviewed (approved):
:shipit:
